teuthology/provision/maas.py

```diff
@@ -0,0 +1,441 @@
+import json
+import logging
+
+from oauthlib.oauth1 import SIGNATURE_PLAINTEXT
+from requests_oauthlib import OAuth1Session
+
+import teuthology.orchestra
+
+from teuthology.config import config
+from teuthology.contextutil import safe_while
+from teuthology.orchestra.opsys import OS
+from teuthology import misc
+
+log = logging.getLogger(__name__)
+
+ubuntu_codenames = {
+    "20.04": "focal",
+    "22.04": "jammy",
+    "24.04": "noble",
+    "25.04": "plucky",
+    "25.10": "questing",
+}
+
+def enabled(warn=False):
+    """Check for required MaaS settings
+
+    :param warn: Whether or not to log a message containing unset parameters
+    :returns: True if they are present; False if they are not
+    """
+    maas_conf = config.get("maas", dict())
+    params = ["api_url", "api_key", "machine_types"]
+    unset = [param for param in params if not maas_conf.get(param)]
+    if unset and warn:
+        log.warning(
+            "MaaS disabled; set the following config options to enable: %s",
+            " ".join(unset),
+        )
+
+    if unset:
+        if not config.get("maas", dict()).get("api_url"):
+            return False
+
+        api_key = config.get("maas", dict()).get("api_key", None)
+        if not api_key:
+            return False
+        if len(api_key.split(":")) < 3:
+            log.warning(
+                "MaaS api_key appears to be malformed; expected format is "
+                "'consumer_key:consumer_token:secret'"
+            )
+            return False
+
+    return True
+
+
+def get_types():
+    """Fetch and parse maas machine_types config
+
+    :returns: The list of MaaS-configured machine types. 
+              An empty list if MAAS is not configured.
+    """
+    if not enabled():
+        return []
+    maas_conf = config.get("maas", dict())
+    types = maas_conf.get("machine_types", "")
+    if not isinstance(types, list):
+        types = types.split(",")
+    return [type_ for type_ in types if type_]
+
+
+def get_session():
+    """Create an OAuth1Session for communicating with the MaaS server"""
+    if not enabled():
+        raise RuntimeError("MAAS is not configured!")
+
+    key, token, secret = config.maas["api_key"].split(":")
+    return OAuth1Session(
+        key,
+        resource_owner_key=token,
+        resource_owner_secret=secret,
+        signature_method=SIGNATURE_PLAINTEXT
+    )
+
+
+class MAAS(object):
+    """Reimage machines with https://maas.io"""
+
+    def __init__(self, name,  os_type=None, os_version=None):
+        """Initialize the MAAS object
+
+        :param name: The fully-qualified domain name of the machine to manage
+        :param os_type: The OS type to deploy (e.g. "ubuntu")
+        :param os_version: The OS version to deploy (e.g. "22.04")
+        """
+        self.session = get_session()
+
+        self.remote = teuthology.orchestra.remote.Remote(
```

Comment:
`teuthology.orchestra.remote` isn't imported explicitly, so this could cause a crash